### ISSUE TITLE
Improve the single CMakeLists.txt file

### ISSUE THREAD

**Issue by jcfreeman2** [DUNE-DAQ/appfwk#34]
_The following is an email I sent Alessandro and Kurt earlier today concerning improvements which can be made to the CMakeLists.txt file. The background is that along with some obvious improvements (don't repeat the same code, e.g.), we're trying to think of CMake functions to write that might make DUNE DAQ developer's lives easier than if they used the standard functions CMake provides._

A few points:

-I agree that it's completely uncontroversial that we want to clean up repeated code and bugs (like an install call that doesn't currently work)

-It also seems to me pretty clear that a standard subproject setup function and daq_add_unit_test function could indeed make life easier

-However, as Reiner said, when you create a set of wrapper functions, you create a new language, which almost by definition will not be documented as well as CMake's functions. Hence wrapper functions must be chosen with care. 

-The dependencies of the executables and libraries in appfwk are sufficiently non-uniform that there aren't any implementations of daq_add_executable and daq_add_library which both simplify the users life by removing interface "knobs" AND perfectly recreate what's currently accomplished in app-fmwk/CMakeLists.txt. E.g., we could say that daq_add_executable and daq_add_library always include a certain set of libraries, but then some of the targets currently defined will get dependencies they don't need. Another drawback is that with this method it's not immediately obvious what all the target's dependencies are. Maybe this is acceptable so users don't need remember to specify dependencies explicitly. Costs and benefits. 

-Orthogonal to the above argument, we could do what tdaq, cetbuildtools, etc. do, which is provide more interface in the wrapper add_library and add_executable functions than their CMake counterparts. E.g., tdaq_add_executable allows you not only to specify source files as dependencies but also libraries. Plus you can also add definitions, etc.
--In one sense it's nice to think you can completely specify an executable in a single line without needing to call functions like target_link_libraries. 
--In another sense, it makes the code harder to read to have these mega-line executable definitions. Even if they wouldn't be that long now, as our codebase swells, they'll become impressively long. You can break them into multiple lines, but then you're virtually back to where you'd have been if you'd just called CMake's add_executable and followed it up with target_link_libraries...without the benefit of knowing that you ONLY had source files, or ONLY had libraries, on a given line. 

Thoughts?

**Comment by jcfreeman2** [comment]
Related: daq-buildtools Issue #3, "Come up with an initial attempt at useful CMake wrapper functions"

**Comment by jcfreeman2** [comment]
Here's an update on ways CMakeLists.txt has been simplified so far. I've added two CMake modules, Findcetlib.cmake and DAQ.cmake, to a new CMake subdirectory in daq-buildtools (currently only available on jcfreeman2/issue3_cmake_wrapper_functions, daq-buildtools's sibling issue to this issue). With those modules, CMakeLists.txt can now call find_package for cetlib (not previously the case), and doesn't need to define point_build_to, since this has been moved into DAQ.cmake. Furthermore, I've added add_unit_test. 

I'll probably change the names of the CMake variables before the pull request, but I'm also working with a model where there's a universal set of libraries and includes, and then a package-specific set of libraries and includes. Right now I have:

Universal libraries: ers, pthread
Universal libraries for executables: the above, plus Boost's program_options
Universal includes: Boost includes

appfwk libraries: appfwk, cetlib, cetlib_except
appfwk includes: none yet
